File: #75_Sort_Colors/Solution.java
/* one pass solution */

public class Solution {
    public void sortColors(int[] nums) {
        if (nums == null || nums.length == 0) return;
        int len = nums.length;
        int rindex = 0; int bindex = len - 1;
        int i = 0;
        while (i <= bindex) {
        	if (nums[i] == 0) {
        		swap(nums, i++, rindex++);
        	}
        	else if (nums[i] == 2) {
        		swap(nums, i, bindex--);
        	}
        	else i++;
        }
    }

    private static void swap(int[] nums, int a, int b) {
    	int tmp = nums[a];
    	nums[a] = nums[b];
    	nums[b] = tmp;
    }

    public static void main(String[] args) {
    	Solution sol = new Solution();
    	int[] t = {1, 2, 0, 1, 2, 0};
    	sol.sortColors(t);
    	for (int g : t) {
    		System.out.print(g + " ");
    	}
    	System.out.println();
    }
}

File: #75_Sort_Colors/Solution.py
class Solution:
    # @param {integer[]} nums
    # @return {void} Do not return anything, modify nums in-place instead.
    def sortColors(self, nums):
        if not nums or len(nums) < 2:
        	return
        i, j = 0, len(nums) - 1
        cur = i
        while cur <= j: #
        	if nums[cur] == 0:
        		self.swap(nums, cur, i)
        		cur += 1
        		i += 1
        	elif nums[cur] == 2:
        		self.swap(nums, cur, j)
        		j -= 1
        	else:
        		cur += 1

    def swap(self, nums, i, j):
    	if not nums:
    		return
    	if i == j:
    		return
    	tmp = nums[j]
    	nums[j] = nums[i]
    	nums[i] = tmp

sol = Solution()
t = [2, 0, 0]
sol.sortColors(t)
print t